File: call.js
import { Client } from "@gradio/client";
import fs from "fs";
import axios from "axios";
import yaml from "js-yaml";
import path from "path";
import os from "os";

// Load advanced options from an external YAML file
function loadYamlConfig(filePath) {
    try {
        const fileContents = fs.readFileSync(filePath, "utf8");
        return yaml.load(fileContents);
    } catch (error) {
        console.error("Error loading YAML file:", error);
        return null;
    }
}

// Path to the external YAML configuration file
const CONFIG_FILE = "advanced_options.yaml";

// Load configuration
const advancedOptions = loadYamlConfig(CONFIG_FILE);

// Define image URLs
const imageUrls = [
    "https://plus.unsplash.com/premium_photo-1664474619075-644dd191935f?fm=jpg&q=60&w=1024",
    "https://plus.unsplash.com/premium_photo-1664474619075-644dd191935f?fm=jpg&q=60&w=1024"
];

// Function to download an image from a URL and save it to a temporary directory
async function downloadImage(url, tempDir) {
    const response = await axios.get(url, { responseType: "arraybuffer" });
    const imageName = path.basename(url);
    const imagePath = path.join(tempDir, imageName);
    fs.writeFileSync(imagePath, response.data);
    return imagePath;
}

// Function to remove temporary files
function removeTempFiles(filePaths) {
    filePaths.forEach(filePath => {
        fs.unlinkSync(filePath);
    });
}

// Function to format the image data for Gradio
function formatImageData(imagePath) {
    return {
        data: fs.readFileSync(imagePath).toString("base64"), // Convert image to base64
        meta: { _type: "gradio.FileData" } // Add the required meta field
    };
}

async function main() {
    // Create a temporary directory
    const tempDir = fs.mkdtempSync(path.join(os.tmpdir(), "gradio-images-"));

    try {
        // Download images to the temporary directory
        const imagePaths = await Promise.all(imageUrls.map(url => downloadImage(url, tempDir)));

        // Format the image data for Gradio
        const formattedImages = imagePaths.map(imagePath => formatImageData(imagePath));

        // Connect to the Gradio client
        const client = await Client.connect("https://005cfef1293a5a9d7e.gradio.live/");

        // Make the prediction request with the formatted image data
        const result = await client.predict("/predict", { 
            images: formattedImages, 		
            lora_name: "Hello!!", 		
            concept_sentence: "Hello!!", 		
            steps: 1000, 		
            lr: 0.0004, 		
            rank: 16, 		
            model_type: "dev", 		
            low_vram: true, 		
            sample_prompts: null, 		
            advanced_options: advancedOptions, 
        });

        console.log(result.data);

    } catch (error) {
        console.error("Error during prediction:", error);
    } finally {
        // Remove temporary files after the request is done
        removeTempFiles(imagePaths);
        // Remove the temporary directory
        fs.rmdirSync(tempDir);
    }
}

// Run the main function
main();

File: call.py
import requests
import yaml
from gradio_client import Client, handle_file
import tempfile
import os

# Load advanced options from an external YAML filex
def load_yaml_config(file_path):
    with open(file_path, "r") as yaml_file:
        return yaml.safe_load(yaml_file)

# Path to the external YAML configuration file
CONFIG_FILE = "advanced_options.yaml"

# Load configuration
advanced_options = load_yaml_config(CONFIG_FILE)

# Initialize the client
client = Client("https://acfca9fcb21bbe293a.gradio.live")

# Define image URLs
image_urls = [
    "https://plus.unsplash.com/premium_photo-1664474619075-644dd191935f?fm=jpg&q=60&w=1024",
    "https://plus.unsplash.com/premium_photo-1664474619075-644dd191935f?fm=jpg&q=60&w=1024",
    "https://plus.unsplash.com/premium_photo-1664474619075-644dd191935f?fm=jpg&q=60&w=1024",
    "https://plus.unsplash.com/premium_photo-1664474619075-644dd191935f?fm=jpg&q=60&w=1024",
    "https://plus.unsplash.com/premium_photo-1664474619075-644dd191935f?fm=jpg&q=60&w=1024",
    "https://plus.unsplash.com/premium_photo-1664474619075-644dd191935f?fm=jpg&q=60&w=1024"
]

# Download images and save them as temporary files
def download_image(url):
    response = requests.get(url, stream=True)
    if response.status_code == 200:
        temp_file = tempfile.NamedTemporaryFile(delete=False, suffix=".jpg")
        with open(temp_file.name, "wb") as file:
            file.write(response.content)
        return temp_file.name  # Return the file path
    else:
        raise Exception(f"Failed to download image from {url}")

# Process image URLs into local temporary files
image_paths = [download_image(url) for url in image_urls]  # Get file paths
image_files = [handle_file(path) for path in image_paths]  # Convert to Gradio format

# Convert YAML data to string format
def yaml_to_string(yaml_data):
    return yaml.dump(yaml_data, default_flow_style=False)

# Send the API request using properly formatted images
result = client.predict(
    images=image_files,  # Pass Gradio-wrapped file references
    lora_name="Hello!!",
    concept_sentence=None,
    steps=1000,
    lr=0.0004,
    rank=16,
    model_type="dev",
    low_vram=False,
    sample_prompts=None,
    advanced_options=yaml_to_string(advanced_options),
    api_name="/predict"
)

# Print the result
print(result)

# Clean up temporary files
for img_path in image_paths:  # Use the original downloaded file paths
    os.remove(img_path)

if __name__ == "__main__":
    pass
